alchemist/alchemist-fxui/src/main/kotlin/it/unibo/alchemist/boundary/monitors/AbstractFXDisplay.kt

```diff
@@ -0,0 +1,287 @@
+    private val effectsCanvas = Canvas()
+    /**
+     * Group dedicated for painting the background.
+     */
+    val background = Group()
+
+    init {
+        firstTime = true
+        setStep(steps)
+        val repaintOnResize = ChangeListener<Number> { _, _, _ ->
+            repaint()
+        }
+        widthProperty().addListener(repaintOnResize)
+        heightProperty().addListener(repaintOnResize)
+        effectsCanvas.isMouseTransparent = true
+        effectsCanvas.widthProperty().bind(widthProperty())
+        effectsCanvas.heightProperty().bind(heightProperty())
+        children.addAll(background, effectsCanvas, interactions.canvases)
+    }
+
+    override fun getViewStatus(): FXOutputMonitor.ViewStatus {
+        return this.viewStatus
+    }
+
+    override fun setViewStatus(viewStatus: FXOutputMonitor.ViewStatus) {
+        this.viewStatus = viewStatus
+    }
+
+    override fun getStep(): Int {
+        return this.step
+    }
+
+    /**
+     * {@inheritDoc}.
+     *
+     * @throws IllegalArgumentException if the step is not bigger than 0
+     */
+    @Throws(IllegalArgumentException::class)
+    override fun setStep(step: Int) {
+        if (step <= 0) {
+            throw IllegalArgumentException("The parameter must be a positive integer")
+        }
+        this.step = step
+    }
+
+    override fun isRealTime(): Boolean {
+        return this.realTime
+    }
+
+    override fun setRealTime(realTime: Boolean) {
+        this.realTime = realTime
+    }
+
+    override fun repaint() {
+        mutex.acquireUninterruptibly()
+        try {
+            if (mayRender.get() && isVisible && !isDisabled) {
+                mayRender.set(false)
+                Platform.runLater {
+                    commandQueue.forEach { it() }
+                    mayRender.set(true)
+                }
+            }
+            interactions.repaint()
+        } catch (e: UninitializedPropertyAccessException) {
+            // wormhole hasn't been initialized
+        }
+        mutex.release()
+    }
+
+    override fun addEffects(effects: Collection<EffectGroup<P>>) {
+        this.effectStack.addAll(effects)
+    }
+
+    override fun addEffectGroup(effects: EffectGroup<P>) {
+        this.effectStack.add(effects)
+    }
+
+    override fun getEffects(): Collection<EffectGroup<P>> {
+        return this.effectStack
+    }
+
+    override fun setEffects(effects: Collection<EffectGroup<P>>) {
+        this.effectStack.clear()
+        this.effectStack.addAll(effects)
+    }
+
+    override fun getKeyboardListener(): KeyboardActionListener = interactions.keyboardListener
+
+    override fun initialized(environment: Environment<T, P>) {
+        stepDone(environment, null, DoubleTime(), 0)
+    }
+
+    override fun stepDone(environment: Environment<T, P>, reaction: Reaction<T>?, time: Time, step: Long) {
+        if (firstTime) {
+            synchronized(this) {
```

Comment:
This reentrant lock is unnecessary... Why is it synchronized?

Reply:
I believe this is necessary so that `init` is called only once ever, even if `firstTime` is true across different threads in the first condition. After the first thread calls `init`, `firstTime` will be false so even when the other threads get to the lock, they won't call `init`.

Reply:
who calls `stepDone`? It's meant to be called solely by the simulation thread. The engine thread calls back this method, so:
1. I don't understand why it needs to deal with concurrency, if there is concurrency in this call we may be doing something wrong
2. In any case, I do not see why make it reentrant, the function is not recursive.
3. The structure resembles [a known Java antipattern](https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java), even though I think it should work here due to the boolean being volatile.
4. Why don't you call `init` in `initialized` instead of calling `stepDone` there and then having to check whether it's the first time?

Reply:
I understand, that makes sense. It doesn't look like there's any concurrency to begin with. I should have looked more into this when I ported this class from Java to Kotlin.
I moved `init` in `initialized` and removed the conditions and the lock from `stepDone`.